File: python/binarySearchTree.py
from linked_list import Linkedlist
from queue import Queue

class Node:
    def __init__(self, l=None, r=None, d=None):
        self.left=l
        self.right=r
        self.data=d

class BinarySearchTree:
    def __init__(self):
        self.root=Node()
        self.__nodeCount=0
    
    def compared(self, n1, n2):
        if(n1>n2): return 1
        elif(n1<n2): return -1
        else: return 0
    
    def contains(self, elem):
        node=self.root
        while(node!=None):
            cmp=self.compared(elem, node.data)
            if(cmp<0):
                node=node.left
            elif(cmp>0):
                node=node.right
            else:
                return True
    
    def __add(self, node: Node, elem):
        if(node==None):
            node=Node(None, None, elem)
        else:
            cmp=self.compared(elem, node.data)
            if(cmp<0):
                node.left=self.__add(node.left, elem)
            else:
                node.right=self.__add(node.right, elem)
        return node

    def add(self, elem):
        if(self.root.data==None):
            self.root.data=elem
            self.__nodeCount+=1
            return
        if(self.contains(elem)): return
        node=self.__add(self.root, elem)
        self.__nodeCount+=1
    
    def __findMin(self, node: Node):
        while(node.left!=None):
            node=node.left
        return node
    
    def __findMax(self, node: Node):
        while(node.right!=None):
            node=node.right
        return node
    
    def __remove(self, node: Node, elem):
        cmp=self.compared(elem, node.data)
        if(cmp<0):
            node.left=self.__remove(node.left, elem)
        elif(cmp>0):
            node.right=self.__remove(node.right, elem)
        
        else:
            if(node.left==None):
                return node.right
            elif(node.right==None):
                return node.left
            
            else:
                tmp=self.__findMax(node.left)
                node.data=tmp.data
                node.left=self.__remove(node.left, tmp.data)
        return node
    
    def size(self):
        return self.__nodeCount

    def remove(self, elem):
        if(self.contains(elem)):
            node=self.__remove(self.root, elem)
            self.__nodeCount-=1

    def hei(self, node: Node):
        if(node ==None): return 0
        return max(self.hei(node.left), self.hei(node.right))+1
    
    def height(self):
        return self.hei(self.root)
    
    def __preorder(self, node: Node):
        if(node==None): return
        print(node.data, end=" ")
        self.__preorder(node.left)
        self.__preorder(node.right)
    
    def preorder(self):
        self.__preorder(self.root)

    def __inorder(self, node: Node):
        if(node==None): return
        self.__inorder(node.left)
        print(node.data, end=" ")
        self.__inorder(node.right)
    
    def inorder(self):
        self.__inorder(self.root)
    
    def __postorder(self, node: Node):
        if(node==None): return
        self.__postorder(node.left)
        self.__postorder(node.right)
        print(node.data, end=" ")
    
    def postorder(self):
        self.__postorder(self.root)

    def levelorder(self):
        q=Queue()
        q.offer(self.root)
        while(q.isEmpty() is False):
            tmp=q.poll()
            print(tmp.data, end=" ")
            if(tmp.left!=None): q.offer(tmp.left)
            if(tmp.right!=None): q.offer(tmp.right)

bst=BinarySearchTree()
for i in range(0, 5):
    bst.add(5-i)
    bst.add(5+i)
print("height:", bst.height())
print("size:", bst.size())
print("preorder traversal:", end=" ")
bst.preorder()
print("\ninorder traversal:", end=" ")
bst.inorder()
print("\npostorder traversal:", end=" ")
bst.postorder()
print("\nLevel order:", end=" ")
bst.levelorder()

    

File: cpp/binary_search_tree.cpp
#include <iostream>
#include <cstdlib>
#include "queue.h"

class BinarySearchTree{
    private:
        class Node{
            public:
                Node *left=NULL, *right=NULL;
                int data=0;
                
                Node(Node* l, Node* r, int val){
                    left=l;
                    right=right;
                    data=val;
                }
        };

        int nodeCount=0;
        Node *root= NULL;

        int compared(int n1, int n2){
            if(n1<n2) return -1;
            else if(n1>n2) return 1;
            else return 0;
        }

        Node* add(Node* node, int elem){
            if(node==NULL)
                node=new Node(NULL, NULL, elem);
            else{
                if(compared(elem, node->data)<0)
                    node->left=add(node->left, elem);
                else
                    node->right=add(node->right, elem);
            }
            return node;
        }

        bool contains(Node* node, int elem){
            if(node==NULL) return false;

            int cmp=compared(elem, node->data);

            if(cmp<0)
                return contains(node->left, elem);
            else if(cmp>0)
                return contains(node->right, elem);
            else
                return true;  
        }

        Node* findMin(Node* node){
            while(node->left!=NULL) node=node->left;
            return node;
        }

        Node* findMax(Node* node){
            while(node->right!=NULL) node=node->right;
            return node;
        }

        Node* remove(Node *node, int elem){
            if(node==NULL) return NULL;

            int cmp=compared(elem, node->data);

            if(cmp<0)
                node->left=remove(node->left, elem);

            else if(cmp>0)
                node->right=remove(node->right, elem);

            else{
                if(node->left==NULL)
                    return node->right;
                else if(node->right==NULL)
                    return node->left;

                else{
                    Node* tmp=findMax(node->left);
                    node->data=tmp->data;
                    node->right=remove(node->right, tmp->data);
                }
            }
            return node;
        }

        int hei(Node* node){
            if(node==NULL) return 0;
            return std::max(hei(node->left), hei(node->right))+1;
        }

        void preorder_t(Node *node){
            if (node==NULL) return;
            std::cout<<node->data<<" ";
            preorder_t(node->left);
            preorder_t(node->right);
        }

        void inorder_t(Node *node){
            if(node==NULL) return;
            inorder_t(node->left);
            std::cout<<node->data<<" ";
            inorder_t(node->right);
        }

        void postorder_t(Node *node){
            if(node==NULL) return;
            postorder_t(node->left);
            postorder_t(node->right);
            std::cout<<node->data<<" ";
        }

        void levelOrder_t(){
            Queue<Node*> q;
            q.offer(root);
            while(!q.isEmpty()){
                Node* node=q.poll();
                if(node->left!=NULL)q.offer(node->left);
                if(node->right!=NULL)q.offer(node->right);
                std::cout<<node->data<<" ";
            }
        }

    public:
        int size(){
            return nodeCount;
        }

        bool isEmpty(){
            return size()==0;
        }

        bool contain(int elem){
            return contains(root, elem);
        }

        bool append(int elem){
            if(contain(elem)) return false;
            root=add(root, elem);
            nodeCount++;
            return true;
        }

        bool remove(int elem){
            if(contain(elem)){
                Node* tmp=remove(root, elem);
                nodeCount--;
                return 1;
            }
            return -1;
        }

        int height(){
            return hei(root);
        }

        void preorder(){
            preorder_t(root);
        }

        void inorder(){
            inorder_t(root);
        }

        void postorder(){
            postorder_t(root);
        }

        void levelorder(){
            levelOrder_t();
        }

};

int main(){
    BinarySearchTree t;
    for(int i=0; i<5; i++){
        t.append(5-i);
        t.append(5+i);
    }
    std::cout<<"height: "<<t.height();
    std::cout<<"\npreorder traversal: ";
    t.preorder();
    std::cout<<"\ninorder traversal: ";
    t.inorder();
    std::cout<<"\npostorder traversal: ";
    t.postorder();
    std::cout<<"\nlevelorder traversal: ";
    t.levelorder();
    return 0;
}